### PR TITLE
dynamic reload cluster authentication info for aggregated API servers

### LINKED ISSUE #82141: delegated authn does not automatically reload when ca bundles change
The delegated authn in the generic apiserver doesn't automatically reload when the default ca bundle changes.  We should update to make this happen transparently.  There is a prototype of how to do this in openshift that we can upstream.

@kubernetes/sig-api-machinery-bugs 
@kubernetes/sig-auth-bugs 
/priority important-soon
/assign

For aggregated apiservers, we created delegated authentication options.  These pull from `kubectl get -n kube-system configmap/extension-apiserver-authentication`.  ConfigMaps are logically dynamic, but the delegated authentication doesn't automatically pick up the changes. This causes drift between intent and reality.

- [x] part 1 - update the core authenticators to allow dynamic changes to x509 verification options
- [x] part 2 - update options to plumb the basic verification functions through
- [x] part 3 - update the apiserver to allow dynamic updates to the tls connection options
- [ ] part 4 - wire up a simple controller to back the dynamic verification functions by an efficient and thread-safe informer/lister

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/kubernetes/kubernetes/issues/82141

Final step to automatically reload cluster authentication.

/kind bug
@kubernetes/sig-api-machinery-bugs 
/priority important-soon

```release-note
apiservers based on k8s.io/apiserver with delegated authn based on cluster authentication will automatically update to new authentication information when the authoritative configmap is updated.
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/cert/server_inspection.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// GetClientCANamesForURL is GetClientCANames against a URL string like we use in kubeconfigs
+func GetClientCANamesForURL(kubeConfigURL string) ([]string, error) {
+	apiserverURL, err := url.Parse(kubeConfigURL)
+	if err != nil {
+		return nil, err
+	}
+	return GetClientCANames(apiserverURL.Host)
+}
+
+// GetServingCertificates returns the x509 certs used by a server.  The serverName is optional for specifying a different
```
Comment (sttts) [inline staging/src/k8s.io/client-go/util/cert/server_inspection.go]: what is the second return value? Why do we need it?

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 2]
```diff
@@ -114,14 +116,41 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
...
+	}
+	clientSigningCert, err := cert.NewSelfSignedCACert(cert.Config{CommonName: "client-ca"}, clientSigningKey)
+	if err != nil {
+		return result, err
+	}
+	clientCACertFile := path.Join(s.SecureServing.ServerCert.CertDirectory, "client-ca.crt")
+	if err := ioutil.WriteFile(clientCACertFile, testutil.EncodeCertPEM(clientSigningCert), 0644); err != nil {
+		return result, err
+	}
+	s.Authentication.ClientCert.ClientCA = clientCACertFile
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/testserver.go]: This plumbing is against the design of the testserver. Use flags to configure. We only do direct options manipulation where it is absolute necessary.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	// preRunCaches are the caches to sync before starting the work of this control loop
+	preRunCaches []cache.InformerSynced
+}
+
+var _ Notifier = &ConfigMapCAContent{}
+var _ CAContentProvider = &ConfigMapCAContent{}
+var _ ControllerRunner = &ConfigMapCAContent{}
+
+// NewDynamicCAContentFromConfigMap returns a CAContentProvider based on a configmap that automatically reloads content
+func NewDynamicCAContentFromConfigMap(purpose, namespace, name, key string, kubeClient kubernetes.Interface) (*ConfigMapCAContent, error) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go]: describe that it is near-realtime via an informer.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go]: what is purpose? godoc

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+// ConfigMapCAContent provies a CAContentProvider that can dynamically react to configmap changes
+// It also fulfills the authenticator interface to provide verifyoptions
+type ConfigMapCAContent struct {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go]: it is actually a controller. Better call it like that such that it is clear it must be run.

File: staging/src/k8s.io/client-go/util/cert/server_inspection.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/client-go/util/cert/server_inspection.go]: > what is the second return value? Why do we need it?

will add to doc.  It's the encoded bytes.  This is commonly used in places where we call it.  Some checks are done against the struct, some against the serialization.

File: cmd/kube-apiserver/app/testing/testserver.go (hunk 2 above)
Comment (deads2k) [reply cmd/kube-apiserver/app/testing/testserver.go]: > This plumbing is against the design of the testserver. Use flags to configure. We only do direct options manipulation where it is absolute necessary.

External callers set flags, but all the internal options are set via options.  See a couple lines above where directories with flags are explicitly set and a few lines down where runtimeconfigs are directly set.  The only flags are set by external callers, but these options just complete "normal" wiring authentication that callers don't need to manipulate.

**Comment by deads2k** [comment]
comments addressed

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 5]
```diff
@@ -114,14 +120,44 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
...
+		proxySigningKey, err := testutil.NewPrivateKey()
+		if err != nil {
+			return result, err
+		}
+		proxySigningCert, err := cert.NewSelfSignedCACert(cert.Config{CommonName: "front-proxy-ca"}, proxySigningKey)
+		if err != nil {
+			return result, err
+		}
+		proxyCACertFile := path.Join(s.SecureServing.ServerCert.CertDirectory, "proxy-ca.crt")
+		if err := ioutil.WriteFile(proxyCACertFile, testutil.EncodeCertPEM(proxySigningCert), 0644); err != nil {
```
Comment (tedyu) [inline cmd/kube-apiserver/app/testing/testserver.go]: Should the proxyCACertFile be deleted if there is error in subsequent handling in this func ?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+// hasCAChanged returns true if the caBundle is different than the current.
+func (c *ConfigMapCAController) hasCAChanged(caBundle []byte) bool {
+	uncastExisting := c.caBundle.Load()
+	if uncastExisting == nil {
+		return true
+	}
+
+	// check to see if we have a change. If the values are the same, do nothing.
+	existing, ok := uncastExisting.(*caBundleAndVerifier)
+	if !ok {
```
Comment (tedyu) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go]: Should c.caBundle be unloaded (cleared) in this case ?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+// CurrentCABundleContent provides ca bundle byte content
+func (c *ConfigMapCAController) CurrentCABundleContent() []byte {
+	uncastObj := c.caBundle.Load()
+	if uncastObj == nil {
+		return nil // this can happen if we've been unable load data from the apiserver for some reason
+	}
+
+	return c.caBundle.Load().(*caBundleAndVerifier).caBundle
```
Comment (tedyu) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/configmap_cafile_content.go]: Probably we should add verification that cast to *caBundleAndVerifier is successful.

File: staging/src/k8s.io/client-go/util/cert/server_inspection.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			}
+			return &tls.Certificate{}, nil
+		},
+	}
+
+	conn, err := tls.Dial("tcp", apiHost, tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
Comment (tedyu) [inline staging/src/k8s.io/client-go/util/cert/server_inspection.go]: It seems we can directly close (without deferring)

File: staging/src/k8s.io/client-go/util/cert/server_inspection.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	// if a name is specified for SNI, set it.
+	if len(serverName) > 0 {
+		tlsConfig.ServerName = serverName
+	}
+
+	conn, err := tls.Dial("tcp", apiHost, tlsConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	conn.Close()
```
Comment (tedyu) [inline staging/src/k8s.io/client-go/util/cert/server_inspection.go]: conn.ConnectionState() depends on completed handshake.

I think we should check the return code from Close() where:
```
	if c.handshakeComplete() {
		alertErr = c.closeNotify()
```
If error is returned from Close(), most likely the assignment on line 77 wouldn't be useful.

Comment (deads2k) [reply staging/src/k8s.io/client-go/util/cert/server_inspection.go]: > conn.ConnectionState() depends on completed handshake.
> 
> I think we should check the return code from Close() where:
> 
> ```
> 	if c.handshakeComplete() {
> 		alertErr = c.closeNotify()
> ```
> 
> If error is returned from Close(), most likely the assignment on line 77 wouldn't be useful.

Yes, this is a reasonable change.  Feel free to get me on slack to review/approve the change.  Please update the rest of the file to be consistent.

Comment (tedyu) [reply staging/src/k8s.io/client-go/util/cert/server_inspection.go]: Created https://github.com/kubernetes/kubernetes/pull/85284

**Pull request merged**